File: Run_Sudoku.java

import java.util.*;
import java.io.*;
import java.nio.file.*;

//Run_Sudoku
//This is the class that a user runs in order to generate or solve
//Sudoku puzzles

//@authors Matt Hino and Christine Chen
//@date 9/14/2016

public class Run_Sudoku {

  //constants
  private static final String CREATE = "create";
  private static final String SOLVE = "solve";
  private static final String YES = "yes";
  private static final String NO = "no";
  private static final boolean ERROR = false;

  //variables
  private final BufferedReader keyboard;
  private final SudokuSolver solver;
  private final Sudoku_Generator generator;

  //constructor
  public Run_Sudoku() {
    this.keyboard = new BufferedReader(new InputStreamReader(System.in));
    this.solver = new SudokuSolver();
    this.generator = new Sudoku_Generator();
  }

  public static void main(String args[]) {
    Run_Sudoku runner = new Run_Sudoku();
    boolean doAgain = false;

    // welcome the user to the program, describe functionality
    runner.welcomeUser();


    do {
      // ask the user if they want to create puzzles or solve a single puzzle
      String userGoal = runner.userGoals();

      /**
      * If the user wants to create puzzles, we give them the option of choosing
      * how many to create and how difficult to make them.
      **/
      if(userGoal.equals(CREATE)) {
        // how many?
        int numberToCreate = runner.getPuzzlesToCreate();
        // what difficulty?
        Sudoku_Generator.Level difficulty = runner.getPuzzleDifficulty();
        // create puzzles
        System.out.println("Creating " + numberToCreate + " puzzles of " + difficulty + " difficulty");

        boolean status = runner.createPuzzles(numberToCreate, difficulty);
        if(status != ERROR) {
          System.out.println("Done creating puzzles! Have fun solving them.");
        } else {
          System.out.println("Whoops! Something went wrong. :(");
        }
      }
      else if(userGoal.equals(SOLVE)) {
        // what is the file name?
        String fileName = runner.getPuzzleToSolve();

        // try and solve the puzzle located at that file location.
        runner.solvePuzzle(fileName);
      } else {
        System.out.println("Whoops! Something went wrong. :(");
      }

      // ask user if they want to exit the program
      String run_again = runner.ask_run_again();
      if(run_again.equals(runner.YES))
      {
        doAgain=true;
      }
      else
      {
        doAgain=false;
      }
    } while (doAgain==true);

    System.out.println("Thanks for using our program! Have a great day");
  }







  // this function will create and populate a folder full of N sudoku puzzles,
  // where 'N' is the number of puzzles to create. it also takes in a difficulty.
  private boolean createPuzzles(int numberPuzzles, Sudoku_Generator.Level difficulty) {
    try {
      // Create separate directories for the puzzles and their solutions, if they don't exist
      File puzzles = new File("puzzles");
      File puzzles_solved = new File("puzzles_solved");

      if(!puzzles.exists()) Files.createDirectory(Paths.get("./puzzles"));
      if(!puzzles_solved.exists()) Files.createDirectory(Paths.get("./puzzles_solved"));

      // Create both the solved and unsolved puzzle, then write them to text Files
      // in appropriate diretories
      Optional<Square[]> puzzle;
      Optional<Square[]> solvedPuzzle;
      for(int i = 0; i < numberPuzzles; ) {
        puzzle = this.generator.get_sudoku_puzzle(difficulty);

        if(puzzle.isPresent()) {
          Square[] present = puzzle.get();
          solvedPuzzle = this.solver.solve_puzzle(present);
          if(solvedPuzzle.isPresent()) {
            long suffix = System.currentTimeMillis();
            SudokuParser.writeSudoku(present, "./puzzles/puzzle_" + i + "_" + suffix + ".txt");
            SudokuParser.writeSudoku(solvedPuzzle.get(), "./puzzles_solved/puzzle_" + i + "_" + suffix + "_solution.txt");
          }
          // increment only if successful in creating a puzzle and it's solution
          ++i;
        }

        // reset the board so we can create a new one next time
        this.generator.refresh_sudoku_grid();
      }


    } catch (IOException ioe) {
      System.out.println("Error creating directories");
      return ERROR;
    }
    return true;
  }

  private void welcomeUser() {
    System.out.println("Welcome to the Sudoku Suite!");
    System.out.println("This program is designed so that you can either create " +
                        "some\nsudoku puzzles or try to solve a predefined one.");
  }

  // This method returns whether or not a user wants to create or solve a puzzle
  private String userGoals() {
    System.out.println("Would you like to 'create' or 'solve'?");
    String result = "";

    try {
      result =  keyboard.readLine();

      while(!result.equals(CREATE) && !result.equals(SOLVE)) {
        handleInvalidResponse(result, "'" + CREATE + "' or '" + SOLVE + "''");
        result = keyboard.readLine();
      }
    }
    catch (Exception e)
    {
      // in case of IOException
    }
    finally
    {
      return result;
    }
  }

  private int getPuzzlesToCreate() {
    System.out.println("How many puzzles would you like to create?");
    String result = "";

    boolean valid = false;
    int numPuzzles = -1;
    try{
      while(!valid) {
        result =  keyboard.readLine();

        try {
          numPuzzles = Integer.parseInt(result);
          if( numPuzzles > 0 ) {
            valid = true;
          } else {
            handleInvalidResponse(result, "Any positive integer");
          }
        } catch (NumberFormatException nfe) {
          handleInvalidResponse(result, "Any positive integer");
        }
      }
    } catch (Exception e) {
      // in case of IOException
    }
    return numPuzzles;
  }

  private Sudoku_Generator.Level getPuzzleDifficulty() {
    System.out.println("Would you like your puzzle difficulty to be \n'easy', " +
                       "'medium', 'difficult', or 'random'?");
    String difficulty = "";
    try {
        difficulty = keyboard.readLine();
        while(!isValidLevelType(difficulty)) {
          handleInvalidResponse(difficulty, "'easy', 'medium', 'difficult', or 'random'");
          difficulty = keyboard.readLine();
        }
    } catch (Exception e) {
      // this would be an IOException. if this happens then ¯\_(ツ)_/¯
    }

    return Sudoku_Generator.Level.valueOf(difficulty.toUpperCase());
  }

  private boolean isValidLevelType(String input) {
    for(Sudoku_Generator.Level level : Sudoku_Generator.Level.values()) {
      if(input.equalsIgnoreCase(level.toString())) {
        return true;
      }
    }
    return false;
  }

  private String getPuzzleToSolve() {
    System.out.println("What is the relative path to the puzzle you are trying to solve?");
    String path = "";
    try {
        path = keyboard.readLine();
    } catch (Exception e) {
      // this would be an IOException. if this happens then ¯\_(ツ)_/¯
    }
    return path;
  }

  // This function is designed to take a fileName passed by the user, parse
  // the sudoku puzzle that is in the file, solve it, and then save it to a new file.
  private void solvePuzzle(String fileName) {
    try {
      Square[] unsolvedPuzzle = SudokuParser.parseSudoku(fileName);
      Optional<Square[]> solvedPuzzle = this.solver.solve_puzzle(unsolvedPuzzle);

      if(solvedPuzzle.isPresent()) {
        String solutionFileName = getSolutionFileName(fileName);
        SudokuParser.writeSudoku(solvedPuzzle.get(), solutionFileName);
        System.out.println("Solved puzzle! See: " + solutionFileName);
      } else {
        System.out.println("No solutions exist for puzzle!");
      }

    } catch (IOException ioe) {
      System.out.println("File not found: " + fileName);
    }
  }

  private String getSolutionFileName(String origName) {
    String removedExtension = origName.substring(0, origName.lastIndexOf('.'));

    return removedExtension + "_solved.txt";
  }

  private static void handleInvalidResponse(String invalidResponse, String validResponses) {
    System.out.println("Sorry, '" + invalidResponse + "' is not a valid response.");
    System.out.println("Please enter one of the following: ");
    System.out.println(validResponses);
  }

  // This method returns whether or not a user wants to run the Sudoku program again
  private String ask_run_again() {
    System.out.println("Would you like to run program again? (Enter 'yes' or 'no')");
    String result = "";

    try {
      result =  keyboard.readLine();

      while(!result.equals(YES) && !result.equals(NO)) {
        handleInvalidResponse(result, "'" + YES + "' or '" + NO + "''");
        result = keyboard.readLine();
      }
    }
    catch (Exception e)
    {
      // in case of IOException
    }
    finally
    {
      return result;
    }
  }

}

File: cpp/Sudoku/Sudoku/Run_Sudoku.cpp
#include <iostream>
#include <stdlib.h>
#include <string>
#include <sstream>
#include <direct.h>
#include <stdio.h>
#include <set>
#include <ctime>
#include <algorithm>

using namespace std;

#include "Run_Sudoku.h"
#include "Sudoku_Generator.h"
#include "Sudoku_Parser.h"
#include "Sudoku_Solver.h"
#include "Square.h"

//constants
const string Run_Sudoku::CREATE = "CREATE";
const string Run_Sudoku::SOLVE = "SOLVE";
const string Run_Sudoku::YES = "YES";
const string Run_Sudoku::NO = "NO";
const bool  Run_Sudoku::OOPS = false;

// constants for comparing what level the user inputted
const string EASY = "EASY";
const string MEDIUM = "MEDIUM";
const string DIFFICULT = "DIFFICULT";
const string RANDOM = "RANDOM";

//variables
Sudoku_Solver* solver;
Sudoku_Generator* generator;

// constructor destructor
Run_Sudoku::Run_Sudoku() {
	solver = new Sudoku_Solver();
	generator = new Sudoku_Generator();
}

Run_Sudoku::~Run_Sudoku() {
	delete solver;
	delete generator;
}

int Run_Sudoku::main() {
	Run_Sudoku * runner = new Run_Sudoku();
	bool do_again = false;

	// welcome user to the program, describe the functionality
	runner->welcome_user();

	do {
		string user_goal = runner->user_goals();

		//	If user wants to create puzzles, give them the following options:
		//	1. how many, 2. what difficulty
		if (user_goal.compare(Run_Sudoku::CREATE) == 0) {
			// get the information from the user
			int num_to_create = runner->get_num_puzzles();
			Level difficulty = runner->get_puzzle_level();
			cout << "Creating " << num_to_create << " puzzles of " << difficulty << " difficulty" << endl;

			// create the puzzles and let the user know the result.
			bool status = runner->create_puzzles(num_to_create, difficulty);
			if (status != Run_Sudoku::OOPS) {
				cout << endl;
				cout << "Done creating puzzles! Have fun solving them!" << endl;
			}
			else {
				cout << endl;
				cout << "Whoops! Something went wrong :(" << endl;
			}
		}
		// ask the user which puzzle they would like to solve, and solve it
		else if (user_goal.compare(Run_Sudoku::SOLVE) == 0) {
			string file_name = runner->get_puzzle_to_solve();
			runner->solve_puzzle(file_name);
		}
		else {
			cout << "Whoops! Something went wrong here. :( " << endl;
		}

		// ask if the user wants to repear the process over again.
		do_again = runner->ask_run_again();
	} while (do_again);



	return 0;
} // main

// displays nice welcome message to user of program.
void Run_Sudoku::welcome_user(void) {
	cout << "Welcome to the Sudoku Suite!" << endl;
	cout << "This program is designed so that you can create ";
	cout << "and solve some" << endl << "sudoku puzzles ";
	cout << "or solved an already created one." << endl << endl;
}

// asks if user wants to create or solve puzzles
string Run_Sudoku::user_goals() {
	cout << "Would you like to 'create' or 'solve'?" << endl;

	string result;
	getline(cin, result);
	
	// transform string to upper case
	transform(result.begin(), result.end(), result.begin(), toupper);
	
	cout << endl;

	// keep asking until the user types 'create' or 'solve'
	while (!(Run_Sudoku::CREATE.compare(result) == 0 ||
			 Run_Sudoku::SOLVE.compare(result) == 0)) {
		handle_invalid_response(result, "'" + Run_Sudoku::CREATE + "' or '" + Run_Sudoku::SOLVE + "'");
		getline(cin, result);
		transform(result.begin(), result.end(), result.begin(), toupper);
		cout << endl;
	}

	return result;
}

// asks user how many puzzles to create
int Run_Sudoku::get_num_puzzles(void) {
	cout << "How many puzzles would you like to create?" << endl;
	string result;

	bool valid = false;
	int num_puzzles = -1;

	// ask the user how many puzzles they want to create
	// if the input is not an integer, handle nicely and ask again.
	while (!valid) {
		getline(cin, result);
		cout << endl;
		try{
			num_puzzles = stoi(result);

			if (num_puzzles > 0) {
				valid = true;
			}

		}
		// handle the appropriate erros
		catch (invalid_argument ia){
			handle_invalid_response(result, "Any positive integer");
		}
		// if they enter a number that's too big
		catch (out_of_range oor) {
			handle_invalid_response(result, "Must be less than" + INT_MAX);
		}
	}
	
	return num_puzzles;
}

// asks user what difficulty puzzle to create
Level Run_Sudoku::get_puzzle_level(void) {
	cout << "Would you like your puzzle difficulty to be" << endl;
	cout << "'easy', 'medium', 'difficult', or 'random'?" << endl;

	string difficulty;
	// until the user types in one of the specified difficulties, continue to ask
	getline(cin, difficulty);
	transform(difficulty.begin(), difficulty.end(), difficulty.begin(), toupper);

	cout << endl;

	Level lvl = Run_Sudoku::string_to_level(difficulty);

	while (lvl == none) {
		handle_invalid_response(difficulty, "'easy', 'medium', 'difficult', or 'random'?");
		getline(cin, difficulty);
		transform(difficulty.begin(), difficulty.end(), difficulty.begin(), toupper);
		lvl = Run_Sudoku::string_to_level(difficulty);
	}
	
	return lvl;
}

// asks user for relative path to puzzle
string Run_Sudoku::get_puzzle_to_solve(void) {
	cout << "What is the relative path to the puzzle you want to solve?" << endl;
	string path;

	getline(cin, path);
	cout << endl;
	return path;
}

// asks user if they want to run the program again
bool Run_Sudoku::ask_run_again(void) {
	cout << endl;
	cout << "Would you like to run the program again?" << endl;
	cout << "Please enter 'yes' or 'no'" << endl;

	string response;
	getline(cin, response);
	transform(response.begin(), response.end(), response.begin(), toupper);

	cout << endl;
	while (YES.compare(response) != 0 && NO.compare(response) != 0) {
		handle_invalid_response(response, "'" + YES + "' or '" + NO + "'");
		getline(cin, response);
		transform(response.begin(), response.end(), response.begin(), toupper);
		cout << endl;
	}

	if (YES.compare(response) == 0) {
		return true;
	}
	else {
		return false;
	}
}

// solves the puzzle at a given path
void Run_Sudoku::solve_puzzle(string file) {
	Square* unsolved_puzzle = Sudoku_Parser::parseSudoku(file);
	if (unsolved_puzzle == nullptr) {
		cout << "Next time please enter a valid file name" << endl;
		return;
	}

	Square* solved = solver->solve_puzzle(unsolved_puzzle);
	if (solved == nullptr) {
		cout << "Unable to solve puzzle!" << endl;
	} else{
		string solution_file = get_solution_filename(file);
		
		Sudoku_Parser::write_sudoku(solved, solution_file);
		cout << "Solved puzzle! See: " << solution_file << endl;
	}
}

bool Run_Sudoku::create_puzzles(int num_puzzles, Level level) {
	const char puzzle_path[] = "./puzzles";
	const char solution_path[] = "./solutions";

	// create directories to hold the puzzles and solutions.
	if (create_folder(puzzle_path) == false) return false;
	if (create_folder(solution_path) == false) return false;

	// now create the puzzles and solutions. save them to 
	// appropriate folders.
	Square* puzzle;
	Square* solved;
	for (int i = 0; i < num_puzzles;) {
		puzzle = generator->get_sudoku_puzzle(level);

		if (puzzle != nullptr) {
			solved = solver->solve_puzzle(puzzle);
			if (solved != nullptr) {
				long now = (long)time(0);

				string puzzle_name = "./puzzles/puzzle_" + to_string(i) + "_" + to_string(now) + ".txt";
				string solution_name = "./solutions/puzzle_" + to_string(i) + "_" + to_string(now) + "_solution_" + ".txt";

				int result_p = Sudoku_Parser::write_sudoku(puzzle, puzzle_name);
				int result_s = Sudoku_Parser::write_sudoku(solved, solution_name);

				if (result_p == 0 && result_s == 0) {
					cout << "Created puzzle " << to_string(i + 1) << " of " << to_string(num_puzzles) << "." << endl;
					i++; // only increment if we can find a solution
				}
			}
		}
	}
	generator->refresh_sudoku_grid();
	return true;
}

// try and create a directory and if there is an err, inspect it
// if it's a path not found err, that's not good. otherwise
// we should be ok.
bool Run_Sudoku::create_folder(const char path[]) {
	if (_mkdir(path) != 0) {
		if (errno == ENOENT) {
			return false;
		}
	}

	return true;
}



// converts a string to the specified lvl. if the lvl is not easy,
// medium, difficult, or random, return none and handle accordingly.
Level Run_Sudoku::string_to_level(string lvl) {
	// upper case
	transform(lvl.begin(), lvl.end(), lvl.begin(), toupper);

	if (EASY.compare(lvl) == 0) {
		return easy;
	}
	else if (MEDIUM.compare(lvl) == 0) {
		return medium;
	}
	else if (DIFFICULT.compare(lvl) == 0) {
		return difficult;
	}
	else if (RANDOM.compare(lvl) == 0) {
		return random;
	}
	else {
		return none;
	}
}

void Run_Sudoku::handle_invalid_response(string invalid_response, string valid_response) {
	cout << "Sorry, '" + invalid_response + "' is not a valid response." << endl;
	cout << "Please enter one of the following: " << endl;
	cout << valid_response << endl;
}

string Run_Sudoku::get_solution_filename(string original) {
	return original + "_solved.txt";
}
